### ISSUE TITLE
Composer wrongly uses cache when new version of @dev exists

### ISSUE THREAD

**Issue by cxj** [composer/composer#5675]
Problem using `@dev` version of Github hosted packages:

`Composer version 1.2.1 2016-09-12 11:27:19`

```sh
$ composer install
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing level-2/transphporm (dev-master 2110304)
    Cloning 2110304c512958797508ea08e0d78c4336494b41 from cache

Writing lock file
Generating autoload files
master
```

```sh
$ git log origin
commit 0ada294c2ecd1698782985a23f3c90c142fab7b4 (refs/remotes/origin/master, refs/remotes/origin/HEAD, refs/remotes/composer/master)
Author: Tom Butler <tom@r.je>
Date:   Tue Sep 13 15:51:36 2016 +0100

    #124 - fixed bug with empty value

commit 2110304c512958797508ea08e0d78c4336494b41 (HEAD -> refs/heads/master)
Author: Richard <solleer@hotmail.com>
Date:   Fri Aug 12 08:17:02 2016 -0400

    Removed unused class properties
```

Note how Composer pulled old version 2110304 from Aug 12 from cache, instead of grabbing version 0ada294c from Github.

(Sorry for the earlier resurrection of an open issue which seemed the same.  I thought duplicated issues were more frowned upon than resurrected ones.)

**Comment by Seldaek** [comment]
```
$ composer show -a level-2/transphporm dev-master
name     : level-2/transphporm
descrip. : A new approach at templating
keywords : template system
versions : dev-master
type     : library
license  : BSD 2-clause "Simplified" License (BSD-2-Clause) (OSI approved) https://spdx.org/licenses/BSD-2-Clause.html#licenseText
source   : [git] https://github.com/Level-2/Transphporm.git 0ada294c2ecd1698782985a23f3c90c142fab7b4
dist     : [zip] https://api.github.com/repos/Level-2/Transphporm/zipball/0ada294c2ecd1698782985a23f3c90c142fab7b4 0ada294c2ecd1698782985a23f3c90c142fab7b4
names    : level-2/transphporm
```

As you can see there, the commit is up to date in the packagist metadata.. I would guess you run install from a lock file but `Updating dependencies (including require-dev)` indicates it's doing an update so you don't have a lock file I assume.

The only option I have left in mind is that you have the commit hardcoded in the composer.json.. can you share the composer.json?

**Comment by Seldaek** [comment]
Also pasting the full output of the `composer update -vvv` would be useful.

**Comment by cxj** [comment]
`composer.json`:

```javascript
{
    "require": {
        "level-2/transphporm":
            "@dev"
    }
}
```

**Comment by cxj** [comment]
I ran `composer update -vvv`, however it succeeded, instead of failing.  Could Packagist just have been out of date when I first ran my install?

**Comment by Seldaek** [comment]
Well.. that is possible but kinda unlikely that the update would have fallen in that half hour between your first post and mine, assuming the package has auto-updates from github enabled. In any case it's kinda hard to address this further if it can't be reproduced :/

**Comment by cxj** [comment]
I was watching the issue in Github, and so saw the commit I needed right away.  I did not have Github in my list of repositories to check, only Packagist.  Slightly later I changed it to include Github, and then it worked.  Then I noticed your examination of the meta data at Packagist, which looks correct (and apparently now works correctly).  Hence, my guess I must have hit my install just before Packagist got updated -- but I admit I don't know what the delay is between Github and Packagist.  It might be only a few seconds, which would tend to make that scenario very improbable.
